### ISSUE TITLE
Docker in Docker

### ISSUE THREAD

**Issue by rosstimson** [drone/drone#380]
Now that Drone supports privileged mode I've been attempting to get a Drone to run Docker inside the container it creates.  I've seen issues #1 and #114 and although similiar they seem more concerned about getting Drone to build a Docker container and push to a registry.  Apologies if you consider this a duplicate to those issues.

I'm using the most popular Docker in Docker container from the original blog post when they first supported doing this:  http://blog.docker.com/2013/09/docker-can-now-run-within-docker/

Everything seems to work fine when play with it manually via Docker however as soon as I attempt to use it with `Drone` I get this common error:

```
Get http:///var/run/docker.sock/v1.12/images/json: dial unix /var/run/docker.sock: no such file or directory
```

## Steps to recreate

Create a test `.drone.yml`

```
image: jpetazzo/dind
env:
  - CI=drone
script:
  - docker -v
  - docker images
```

Run `drone --privileged build`

### Output

```
[DRONE] starting build 
$ docker -v
Docker version 1.0.1, build 990021a
$ docker images
2014/07/20 14:35:21 Get http:///var/run/docker.sock/v1.12/images/json: dial unix /var/run/docker.sock: no such file or directory
```

I think this might be due to the 'magic wrapper' script which is set as the `CMD` in the `dind` Dockerfile not being run when runing via the `drone` command but it is used when you run the container directly with `docker` command.

Any ideas on getting this working, has anyone managed to get Docker working within Drone yet?

**Comment by exocom** [comment]
Thanks for the hard work on drone! I've put together a video on some ideas on how the pipeline could work to support docker images and docker cache so that you wouldn't need to do Docker in Docker.  The video is a little long, so just put it on 2x speed.

https://www.youtube.com/watch?v=qkzswVOPgSE

Let me know what you guys think, I'll loved to see drone be able to do more with docker. Anyone else have an existing workflow they can share? I read on another issue about 2 approaches and I think this one is an off shoot of using drone to talk to host docker.

Here is the example of what the yaml file could look like with a pipeline:

```
dockerfile:
  - name: Dockerfile
    env:
      - PORT=2375
      - NODE_ENV=dev
    pipeline:
     - startup
     - test:
       args:
         entrypoint: /bin/bash
       services:
         - mysql:latest
         - mongo:latest
         - custom-npm-cache:latest
       script:
         - npm test
       helpers:
         - headless-chrome:
         args:
           volume: /usr/src/app:/root/test
         script:
          - karma start
     - complie
       args:
         entrypoint: /bin/bash
         mount-folder: /usr/src/app
       script:
         - gulp build
  - name: dist.Dockerfile
    env:
      - NODE_ENV=prod
    pipline:
      - startup
      - test:
        args:
          entrypoint: /bin/bash
        services:
          - mysql:latest
          - mongo:latest
          - custom-npm-cache:latest
        script:
          - npm test
deployment:
  pipeline:
    - docker-registry:
      image: cutsom-ssh-box
      mount-folder: /docker-images
      script:
        - ssh #connect to registry
        - cd /docker-images
        - docker push myUser/myImage
    - etcd:
      image: ubuntu:latest
      script:
        - etcdctl set /image/hash true
    - email:
      image: myCompany/email
      script:
        - . sendmail.sh
notify:
  email:
    recipients:
          - brad@drone.io
          - burke@drone.io

```

**Comment by bradrydzewski** [comment]
fixed in 0.4 branch (upcoming release) through use of the [drone-plugins/drone-docker](https://github.com/drone-plugins/drone-docker) plugin. It runs Docker-in-Docker in privileged mode, without running your build container in privileged mode

also have the ability to mount volumes as part of the yaml, in case you would prefer to mount the Docker unix socket as a volume
